# Data Profiling

#### dataset-of-00s.csv

In [ ]:
!pip install -U -q pydrive

In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
import pydrive as pyd
import io
from pandas_profiling import ProfileReport
from google.colab import drive
from google.colab import files

uploaded = files.upload()

Saving dataset-of-00s.csv to dataset-of-00s.csv


In [ ]:
df = pd.read_csv(io.BytesIO(uploaded['dataset-of-00s.csv']))
df.head()

track             artist  \
0                 Lucky Man  Montgomery Gentry   
1            On The Hotline       Pretty Ricky   
2        Clouds Of Dementia         Candlemass   
3  Heavy Metal, Raise Hell!      Zwartketterij   
4           I Got A Feelin'   Billy Currington   

                                    uri  danceability  energy  key  loudness  \
0  spotify:track:4GiXBCUF7H6YfNQsnBRIzl         0.578   0.471    4    -7.270   
1  spotify:track:1zyqZONW985Cs4osz9wlsu         0.704   0.854   10    -5.477   
2  spotify:track:6cHZf7RbxXCKwEkgAZT4mY         0.162   0.836    9    -3.009   
3  spotify:track:2IjBPp2vMeX7LggzRN3iSX         0.188   0.994    4    -3.745   
4  spotify:track:1tF370eYXUcWwkIvaq3IGz         0.630   0.764    2    -4.353   

   mode  speechiness  acousticness  instrumentalness  liveness  valence  \
0     1       0.0289      0.368000           0.00000     0.159    0.532   
1     0       0.1830      0.018500           0.00000     0.148    0.688   
2     1       0.0473      0.000111           0.00457     0.174    0.300   
3     1       0.1660      0.000007           0.07840     0.192    0.333   
4     1       0.0275      0.363000           0.00000     0.125    0.631   

     tempo  duration_ms  time_signature  chorus_hit  sections  target  
0  133.061       196707               4    30.88059        13       1  
1   92.988       242587               4    41.51106        10       1  
2   86.964       338893               4    65.32887        13       0  
3  148.440       255667               4    58.59528         9       0  
4  112.098       193760               4    22.62384        10       1

For this dataset, we will check for data type and format, descriptive statistics, duplicate values, missing values, and errors.

In [ ]:
df.sort_values(by="artist")

track  \
2134                              Canadian Idiot   
1250                               White & Nerdy   
4976                                     Breathe   
1408                                    Blackout   
569                           Blood in the Water   
...                                          ...   
1394                       I Got It From My Mama   
5841                         Jég ( Metál Remix )   
5400  0952 - Remastered 10th Anniversary Edition   
4244                                Erla's Waltz   
535                                       Ljósið   

                                          artist  \
2134                         "Weird Al" Yankovic   
1250                         "Weird Al" Yankovic   
4976  'In The Heights' Original Broadway Company   
1408  'In The Heights' Original Broadway Company   
569                    'Legally Blonde' Ensemble   
...                                          ...   
1394                                   will.i.am   
5841                                      Árnyak   
5400                              Ólafur Arnalds   
4244                              Ólafur Arnalds   
535                               Ólafur Arnalds   

                                       uri  danceability  energy  key  \
2134  spotify:track:1gloYGAZI6eHp6MEPjLuL3         0.543   0.697    8   
1250  spotify:track:60R2v9lheAu3lwZwAFxMZK         0.791   0.613    1   
4976  spotify:track:6OSTrvIX8JXxRxFT4C9y0J         0.397   0.242   10   
1408  spotify:track:7ahhev2IIbivmkRqLjm7g3         0.587   0.667    7   
569   spotify:track:2ewb4VJoSMG1FuSkFKoMgO         0.578   0.293    1   
...                                    ...           ...     ...  ...   
1394  spotify:track:2q2pZHWEQe6w9PgACCzMtC         0.888   0.777    6   
5841  spotify:track:6KjzDfZNazdnPzhQunA7mZ         0.565   0.541    8   
5400  spotify:track:1wmgnPAD7a8mO42WGfn9dq         0.221   0.093    3   
4244  spotify:track:5Ag8TTr1IdnqqLWJjVwpsl         0.443   0.110    6   
535   spotify:track:1tlJp4XADa7tYkPmgX894R         0.482   0.149    5   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
2134    -9.211     1       0.0612       0.00206          0.000002    0.3430   
1250   -11.628     0       0.0763       0.09860          0.000000    0.0765   
4976   -10.754     1       0.0663       0.84800          0.000000    0.1060   
1408    -5.491     0       0.0497       0.44000          0.000000    0.0778   
569    -12.594     1       0.4790       0.64700          0.000000    0.3340   
...        ...   ...          ...           ...               ...       ...   
1394    -6.262     1       0.0505       0.09610          0.000031    0.0514   
5841   -12.178     1       0.0307       0.00629          0.677000    0.1180   
5400   -23.238     1       0.0367       0.93100          0.957000    0.0757   
4244   -18.593     0       0.0955       0.99600          0.959000    0.0882   
535    -11.910     1       0.0288       0.99200          0.399000    0.1010   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
2134    0.861  185.978       143040               4    49.44141         7   
1250    0.896  143.017       170640               4    73.98449         7   
4976    0.144  171.747       244067               3    24.21145        11   
1408    0.453   99.727       237800               4    31.90800        10   
569     0.510  139.609       289960               4    48.63488        11   
...       ...      ...          ...             ...         ...       ...   
1394    0.876  118.998       241520               4    24.27560        12   
5841    0.242  129.825       339027               4    52.78284        13   
5400    0.127  102.960       184169               4    24.21138        12   
4244    0.152   66.481       134714               4    33.42967         7   
535     0.281  146.472       209682               4    77.89839         9   

      target  
2134       1  
1250       1

To find out our data types:

In [ ]:
df.dtypes

track                object
artist               object
uri                  object
danceability        float64
energy              float64
key                   int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
time_signature        int64
chorus_hit          float64
sections              int64
target                int64
dtype: object

To find out some descriptive statistics of our dataset:

In [ ]:
df.describe()

danceability       energy          key     loudness         mode  \
count   5872.000000  5872.000000  5872.000000  5872.000000  5872.000000   
mean       0.542931     0.694511     5.276396    -7.449258     0.645095   
std        0.190003     0.237792     3.564153     5.102543     0.478525   
min        0.058800     0.000348     0.000000   -47.327000     0.000000   
25%        0.416000     0.567000     2.000000    -8.315000     0.000000   
50%        0.556000     0.744000     5.000000    -6.041500     1.000000   
75%        0.681000     0.885000     8.000000    -4.562500     1.000000   
max        0.986000     0.999000    11.000000     1.137000     1.000000   

       speechiness  acousticness  instrumentalness     liveness      valence  \
count  5872.000000   5872.000000       5872.000000  5872.000000  5872.000000   
mean      0.092360      0.214374          0.150917     0.196141     0.482262   
std       0.094997      0.296511          0.301452     0.161965     0.254567   
min       0.022400      0.000000          0.000000     0.019300     0.000000   
25%       0.036175      0.004553          0.000000     0.093700     0.278000   
50%       0.052700      0.060300          0.000022     0.131000     0.486000   
75%       0.107000      0.312000          0.047275     0.263000     0.687000   
max       0.950000      0.996000          0.998000     0.987000     0.982000   

             tempo   duration_ms  time_signature   chorus_hit     sections  \
count  5872.000000  5.872000e+03     5872.000000  5872.000000  5872.000000   
mean    121.613018  2.581706e+05        3.913999    40.734295    11.056880   
std      30.179885  1.395341e+05        0.404280    20.245637     5.758186   
min      46.755000  1.592000e+04        0.000000     0.000000     1.000000   
25%      96.984000  2.068130e+05        4.000000    27.507507     9.000000   
50%     119.999000  2.380065e+05        4.000000    36.037160    10.000000   
75%     141.495250  2.791600e+05        4.000000    47.889570    12.000000   
max     213.233000  4.170227e+06        5.000000   262.615400   169.000000   

            target  
count  5872.000000  
mean      0.500000  
std       0.500043  
min       0.000000  
25%       0.000000  
50%       0.500000  
75%       1.000000  
max       1.000000

The duration column for each song is presented in milliseconds. It would be interesting to convert these values into minutes and seconds instead for an easier readability.

To check if our dataframe has any missing values:

In [ ]:
df.isnull().sum()

track               0
artist              0
uri                 0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
duration_ms         0
time_signature      0
chorus_hit          0
sections            0
target              0
dtype: int64

We can also find out how many tracks each artist has in the dataset collected:

In [ ]:
artist_counts = df["artist"].value_counts()
artist_counts.sort_values(ascending=False)

Toby Keith                        27
Rascal Flatts                     24
Tim McGraw                        24
Kenny Chesney                     23
Iron Maiden                       23
                                  ..
Ciara Featuring Ludacris           1
Lisbon Improvisation Players       1
Kid Rock Featuring Sheryl Crow     1
Twista Featuring Pitbull           1
Wynonna                            1
Name: artist, Length: 2920, dtype: int64

We have 2920 unique artists in which Toby Keith shows up the most with 27 different tracks.

Alternatively, we can check if there is any song duplicates:

In [ ]:
tracks_counts = df["track"].value_counts()
tracks_counts[tracks_counts > 1]

Breathe           5
Beautiful         4
The One           4
Forever           4
Closer            4
                 ..
South Side        2
Magnificent       2
O-Block Tempel    2
Lucky             2
The Messenger     2
Name: track, Length: 218, dtype: int64

In [ ]:
df[df["track"]=="Breathe"]

track                                      artist  \
2471  Breathe       Taylor Swift Featuring Colbie Caillat   
4214  Breathe                             Michelle Branch   
4976  Breathe  'In The Heights' Original Broadway Company   
5186  Breathe                                    Fabolous   
5858  Breathe                                  Faith Hill   

                                       uri  danceability  energy  key  \
2471  spotify:track:6S9G7J3EB1ZY0rrZPwsBg5         0.498   0.486    1   
4214  spotify:track:635mzxhITS1xY3AVSLKgyD         0.644   0.753    6   
4976  spotify:track:6OSTrvIX8JXxRxFT4C9y0J         0.397   0.242   10   
5186  spotify:track:7fuoqve1SOEBmhc2PoQTF0         0.713   0.910   10   
5858  spotify:track:3y4LxiYMgDl4RethdzpmNe         0.529   0.496    7   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
2471    -7.384     1       0.0273        0.3730           0.00000     0.113   
4214    -4.296     1       0.0391        0.0279           0.00227     0.108   
4976   -10.754     1       0.0663        0.8480           0.00000     0.106   
5186    -5.586     0       0.2820        0.3400           0.00000     0.304   
5858    -9.007     1       0.0290        0.1730           0.00000     0.251   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
2471    0.170  147.952       263987               4    55.20851        10   
4214    0.209  117.148       211773               4    25.66600         9   
4976    0.144  171.747       244067               3    24.21145        11   
5186    0.510   83.028       268960               4    18.81209        12   
5858    0.278  136.859       250547               4    43.09482        12   

      target  
2471       1  
4214       1  
4976       0  
5186       1  
5858       1

"Breathe" is not duplicated since the songs are from different artists.

In [ ]:
df[df["track"]=="The Messenger"]

track   artist                                   uri  \
2499  The Messenger  Turisas  spotify:track:6OcuvCJr6ohpUopzX0HDIL   
3465  The Messenger  Turisas  spotify:track:5yvGEIQWLQBME5lNOY7CMG   

      danceability  energy  key  loudness  mode  speechiness  acousticness  \
2499         0.223   0.889    7     -5.76     1       0.0694      0.000097   
3465         0.127   0.879    7     -5.77     1       0.0857      0.000111   

      instrumentalness  liveness  valence    tempo  duration_ms  \
2499             0.487     0.124    0.282  119.613       282333   
3465             0.422     0.126    0.257   79.875       282333   

      time_signature  chorus_hit  sections  target  
2499               4    24.55661        10       0  
3465               4    24.30747        10       0

"The Messenger" by Turisas seems to be duplicated since both tracks have the same duration and share some other metrics. Investigating more, it seems that they are the same song released within two different albums.

Further investigation is neccessary to check for other duplicated songs in this dataset:

In [ ]:
list_repeated_songs = tracks_counts[tracks_counts > 1].index.to_list()
sorted(list_repeated_songs)


['4 Minutes',
 'Addicted',
 'Again',
 'All Falls Down',
 'All Summer Long',
 'Alone',
 'Already Gone',
 'Always',
 'Amazing',
 'Among Ancestors',
 'Amor Eterno',
 "And I Am Telling You I'm Not Going",
 'Angel',
 'Angels',
 'Anything',
 'Apologize',
 'Apparition',
 'Back At One',
 'Bad Day',
 'Bani Bani - From "Main Prem Ki Diwani Hoon"',
 'Beautiful',
 'Because Of You',
 'Beep',
 'Believe',
 'Best I Ever Had',
 'Boston',
 'Breathe',
 'Bright Lights',
 'Broken',
 'Burn',
 "Burnin' Up",
 'By Your Side',
 'Candy',
 'Closer',
 'Come Back To Me',
 'Coming Back Around - Domu Re-Beat Mix',
 'Crazy',
 'Crush',
 'Daft Punk Is Playing at My House - Soulwax Shibuya Mix',
 'Damaged',
 'Dance Tonight',
 'Dance With Me',
 'Days Go By',
 "Don't Know Why",
 "Don't Tell Me",
 'Dreams',
 'Drive',
 'Electricity',
 'Energy',
 'Fallout (Of Our Being)',
 'Female Buddha Taras Mantra - Tibetan Chant plus Urban Groove',
 'Forever',
 'Forgotten Scriptures',
 'Get Back',
 'Get Up',
 'Gimme Danger',
 'Gimmie Love

In [ ]:
repeated_df = df[df["track"].isin(list_repeated_songs)].sort_values(by="track")
repeated_df

track                                           artist  \
1313          4 Minutes                                            Avant   
1750          4 Minutes  Madonna Featuring Justin Timberlake & Timbaland   
3846           Addicted                                      Simple Plan   
5816           Addicted                                      Saving Abel   
208               Again                                      Faith Evans   
...                 ...                                              ...   
4688    You Raise Me Up                                    Secret Garden   
3105    You Raise Me Up                                      Josh Groban   
2612    You Raise Me Up       Josh Groban & The African Children's Choir   
4432  You Rock My World                                  Michael Jackson   
98    You Rock My World                                      Java Gospel   

                                       uri  danceability  energy  key  \
1313  spotify:track:3g6SuyjGaGyAXmjvdNI1IU         0.828   0.600   10   
1750  spotify:track:1oHClQEgDmmbcEx12Kc5nZ         0.753   0.931    2   
3846  spotify:track:1B8MHa0NveCF1ViBEOZWO5         0.522   0.751    1   
5816  spotify:track:3pxCXv0NACYfYnLjFcQW5D         0.512   0.865    7   
208   spotify:track:17Yws6NV4W7lAGPtq4UPIT         0.489   0.763    7   
...                                    ...           ...     ...  ...   
4688  spotify:track:1J5jCQ4GCR3QK5TNDYVHze         0.263   0.151    0   
3105  spotify:track:4TbNLKRLKlxZDlS0pu7Lsy         0.250   0.136   10   
2612  spotify:track:4TbNLKRLKlxZDlS0pu7Lsy         0.250   0.136   10   
4432  spotify:track:1Lhwn4PqeGpM4LTVUowW76         0.854   0.673    4   
98    spotify:track:00nQhTtc8FxgmB0NWypSE0         0.662   0.854    1   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
1313    -5.429     1       0.1620      0.094200          0.000000    0.1260   
1750    -4.922     1       0.0652      0.009940          0.006960    0.2340   
3846    -5.460     1       0.0435      0.000491          0.000002    0.1580   
5816    -4.146     1       0.0338      0.000821          0.000000    0.0982   
208     -6.077     1       0.2400      0.015800          0.000000    0.4200   
...        ...   ...          ...           ...               ...       ...   
4688   -13.078     1       0.0386      0.917000          0.006580    0.1030   
3105   -12.162     1       0.0370      0.673000          0.000157    0.1070   
2612   -12.162     1       0.0370      0.673000          0.000157    0.1070   
4432    -3.132     0       0.1850      0.038000          0.000227    0.2550   
98      -5.518     1       0.0434      0.133000          0.000000    0.3330   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
1313   0.6550  119.969       240320               4    22.92460        11   
1750   0.7670  113.029       189693               4    87.62425         8   
3846   0.6050  139.510       232067               3    15.96437        13   
5816   0.5290  138.017       222827               4    71.29522         7   
208    0.3620  173.822       198787               4    41.44659        11   
...       ...      ...          ...             ...         ...       ...   
4688   0.0698  134.204       304040               4    58.84595        13   
3105   0.0981  118.086       292333               4    28.80315        12   
2612   0.0981  118.086       292333               4    28.80315        12   
4432   0.9550   95.000       337733               4    32.92256         9   
98     0.9150  121.961       245947               4    40.45327        11   

      target  
1313       1  
1750       1  
3846       1  
5816       1  
208        1  
...      ...  
4688       0  
3105       1  
2612       1  
4432       1  
98         0  

[471 rows x 19 columns]

In [ ]:
repeated_df.groupby(["artist", "track"])["artist"].count()[repeated_df.groupby(["artist", "track"])["artist"].count() > 1]

artist                           track                                                        
Al Storm                         Surrender - Hardcore Adrenaline Mix                              2
Amistades Peligrosas             Lágrimas de metal                                                2
Apocalyptica                     I Don't Care (feat. Adam Gontier)                                2
                                 Path                                                             2
Bobina                           Slow - Grad Big Room Mix                                         2
Candlemass                       Hammer Of Doom                                                   2
Charles Mingus                   Wednesday Night Prayer Meeting - Alternative Take                2
Collie Buddz                     Gimmie Love                                                      2
Digger Dance                     O-Block Tempel                                                   2
Disgo

These should be all the duplicated songs.

In [ ]:
df[df["track"]=="Here Comes the Sun"]

track    artist                                   uri  \
2616  Here Comes the Sun  Yo-Yo Ma  spotify:track:3Kyj1MN9XwnxT8wNDchLl5   
4827  Here Comes the Sun  Yo-Yo Ma  spotify:track:3FO4zWTQ484C2P7JCddeH0   

      danceability  energy  key  loudness  mode  speechiness  acousticness  \
2616         0.366   0.164    5   -14.356     1       0.0386         0.915   
4827         0.366   0.164    5   -14.356     1       0.0386         0.915   

      instrumentalness  liveness  valence    tempo  duration_ms  \
2616          0.000749     0.108    0.463  114.848       170387   
4827          0.000749     0.108    0.463  114.848       170387   

      time_signature  chorus_hit  sections  target  
2616               5    27.78748        10       0  
4827               5    27.78748        10       0

Analyzing "Here Comes the Sun" by Yo-Yo Ma, both songs have the same metrics with a different uri. The two uris represent the same track in the same album.

In [ ]:
duplicateTracks = df[df.duplicated(['track', 'artist'], keep=False)].sort_values(by='track')
duplicateTracks

track             artist  \
4128                                    Among Ancestors            Turisas   
3033                                    Among Ancestors            Turisas   
4540        Bani Bani - From "Main Prem Ki Diwani Hoon"      K. S. Chithra   
3417        Bani Bani - From "Main Prem Ki Diwani Hoon"      K. S. Chithra   
59                Coming Back Around - Domu Re-Beat Mix               Domu   
...                                                 ...                ...   
4953                                  We Are The People  Empire of the Sun   
249    We Will Glorify - House Of Worship Album Version        Twila Paris   
4498   We Will Glorify - House Of Worship Album Version        Twila Paris   
151   Wednesday Night Prayer Meeting - Alternative Take     Charles Mingus   
5363  Wednesday Night Prayer Meeting - Alternative Take     Charles Mingus   

                                       uri  danceability  energy  key  \
4128  spotify:track:5jtlhT1Mq1lTUXYqG7DK1j         0.243   0.843    2   
3033  spotify:track:6i8B7ZVXJn0zIbGbSlR0cu         0.237   0.847    6   
4540  spotify:track:0nCrzqqXQB7tC3GwZTuLIO         0.739   0.575    5   
3417  spotify:track:3vlaZLrpL8pbbZKsDTzkyE         0.739   0.575    5   
59    spotify:track:3ux9LeNRAgENFLcI7ZFebb         0.364   0.866    1   
...                                    ...           ...     ...  ...   
4953  spotify:track:57RHMnLMQx8Qz5V6c0E8dF         0.666   0.785    4   
249   spotify:track:0V3aZTzeoDK6whbZwsKX6w         0.468   0.554    9   
4498  spotify:track:5q4oEZ9e0dJAejnTA6GHo6         0.479   0.565    9   
151   spotify:track:42IqaW4VDhFomQMojirykk         0.377   0.453    5   
5363  spotify:track:13dsMonoAHC8iHNmzKccLr         0.434   0.456    5   

      loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
4128    -7.558     1       0.2850       0.00401          0.492000    0.0786   
3033    -7.622     0       0.3020       0.00388          0.455000    0.0740   
4540   -12.586     1       0.0780       0.44500          0.000004    0.3610   
3417   -12.586     1       0.0780       0.44500          0.000004    0.3610   
59      -6.798     1       0.1390       0.01340          0.006180    0.1090   
...        ...   ...          ...           ...               ...       ...   
4953    -5.370     0       0.0308       0.09810          0.019800    0.4340   
249     -5.580     1       0.0315       0.74400          0.000000    0.0996   
4498    -6.190     1       0.0316       0.64500          0.000000    0.1070   
151    -14.669     0       0.0636       0.45900          0.002680    0.0921   
5363   -12.454     0       0.0666       0.67700          0.014900    0.1000   

      valence    tempo  duration_ms  time_signature  chorus_hit  sections  \
4128    0.182   87.170       316613               3    78.30956        12   
3033    0.237  169.483       316613               4    69.96880        13   
4540    0.751  112.081       422067               4    52.73578        13   
3417    0.751  112.081       422067               4    52.73578        13   
59      0.963  189.091       374773               3    29.83122        19   
...       ...      ...          ...             ...         ...       ...   
4953    0.549  122.973       267355               4    44.49845        10   
249     0.175  135.903       193427               3    44.34413         6   
4498    0.166  135.927       192640               3    46.04194         6   
151     0.607  126.930       341333               3    35.14482        15   
5363    0.507  127.403       338640               3    34.32397        16   

      target  
4128       0  
3033       0  
4540       0  
3417       0  
59         0  
...      ...  
4953       0  
249        0  
4498       0  
151        0  
5363       0  

[88 rows x 19 columns]

In [ ]:
duplicateTracks.head(22)

track  \
4128                                    Among Ancestors   
3033                                    Among Ancestors   
4540        Bani Bani - From "Main Prem Ki Diwani Hoon"   
3417        Bani Bani - From "Main Prem Ki Diwani Hoon"   
59                Coming Back Around - Domu Re-Beat Mix   
3194              Coming Back Around - Domu Re-Beat Mix   
35    Daft Punk Is Playing at My House - Soulwax Shi...   
3007  Daft Punk Is Playing at My House - Soulwax Shi...   
123                              Fallout (Of Our Being)   
2147                             Fallout (Of Our Being)   
3775  Female Buddha Taras Mantra - Tibetan Chant plu...   
201   Female Buddha Taras Mantra - Tibetan Chant plu...   
1647                               Forgotten Scriptures   
778                                Forgotten Scriptures   
3262                                       Gimme Danger   
3013                                       Gimme Danger   
2262                                        Gimmie Love   
235                                         Gimmie Love   
2892                                     Hammer Of Doom   
3430                                     Hammer Of Doom   
1231                                Heavy Metal Drummer   
1953                                Heavy Metal Drummer   

                               artist                                   uri  \
4128                          Turisas  spotify:track:5jtlhT1Mq1lTUXYqG7DK1j   
3033                          Turisas  spotify:track:6i8B7ZVXJn0zIbGbSlR0cu   
4540                    K. S. Chithra  spotify:track:0nCrzqqXQB7tC3GwZTuLIO   
3417                    K. S. Chithra  spotify:track:3vlaZLrpL8pbbZKsDTzkyE   
59                               Domu  spotify:track:3ux9LeNRAgENFLcI7ZFebb   
3194                             Domu  spotify:track:20nhrlq4tB7OS8GaXrHTrG   
35                    LCD Soundsystem  spotify:track:7d7l6Qnbnqv9nkuHrh6p1N   
3007                  LCD Soundsystem  spotify:track:3Xi96h0cAiDJqVAQsbPLHF   
123                            Nausea  spotify:track:6SO0RfXtb6kgYO17lNwKn8   
2147                           Nausea  spotify:track:2AyIFQ6caNfTpVLef1xqWr   
3775  Lama Surya Das & Steven Halpern  spotify:track:3SvZOri3IfrcrlOpbqp8zx   
201   Lama Surya Das & Steven Halpern  spotify:track:4jltSJWHhH9IBq980mhfTg   
1647                         Disgorge  spotify:track:3p8KgeEYByP6F3xAp7ylKi   
778                          Disgorge  spotify:track:1fTG9HSw7saP8GIScPkGt2   
3262                      The Stooges  spotify:track:0yOcVe3GzHnKT2bdlVb8jR   
3013                      The Stooges  spotify:track:3jmVkbF3KSFG8xDzvirYQg   
2262                     Collie Buddz  spotify:track:7BXJKrUhH5Z61kMf4LDIMr   
235                      Collie Buddz  spotify:track:6UnR7DiYC4uGQW6YxkVcWT   
2892                       Candlemass  spotify:track:0rwfwTWMlH7i2vEhsqpwNz   
3430                       Candlemass  spotify:track:7qkpZGn8vx2kOXnNXou5FU   
1231                            Wilco  spotify:track:4xmeLj3uVcoAwgDtwM4MbG   
1953                            Wilco  spotify:track:4Z3LaXjxO8tdYX1ocJiLtl   

      danceability  energy  key  loudness  mode  speechiness  acousticness  \
4128         0.243   0.843    2    -7.558     1       0.2850      0.004010   
3033         0.237   0.847    6    -7.622     0       0.3020      0.003880   
4540         0.739   0.575    5   -12.586     1       0.0780      0.445000   
3417         0.739   0.575    5   -12.586     1       0.0780      0.445000   
59           0.364   0.866    1    -6.798     1       0.1390      0.013400   
3194         0.364   0.866    1    -6.798     1       0.1390      0.013400   
35           0.658   0.954    9    -5.376     1       0.0792      0.000891   
3007         0.658   0.954    9    -5.376     1       0.0792      0.000891   
123          0.308   0.935    2    -7.309     0       0.1240      0.002540   
2147         0.292   0.964    2    -5.272     1       0.0935      0.000003   
3775         0.5

In [ ]:
#looking at irregularities
#1 track with loudness > 1 (range for this variable is 0.0-1.0)
df.loc[df['loudness']>1.0]

track     artist                                   uri  \
642  Dance of Lust  Freakshow  spotify:track:1ZlCP70vdRmQI4qMvPpXIS   

     danceability  energy  key  loudness  mode  speechiness  acousticness  \
642         0.452   0.983    2     1.137     0       0.0694       0.00136   

     instrumentalness  liveness  valence    tempo  duration_ms  \
642          0.000192     0.133    0.556  148.829       216880   

     time_signature  chorus_hit  sections  target  
642               4    52.53548        13       0

In [ ]:
#Zero values on chorus_hit...seems to be tracks with just sounds or no distinct "sections"
df.loc[df['chorus_hit']==0]

track  \
145                                 Full Metal Swimsuit   
255   Interzone, Pt. 1: Part I: Why don't you have a...   
631                                                 Ten   
666                                    Oxygen Corrosion   
1595                                     Concrete Waves   
2126  10,000 Angry Hardcore Kids - Wormwood Album Ve...   
2222                               Under an Ancient Oak   
3431          Subliminal Heavy Metal - Original Version   
3621                                Let's Rock Shall We   
4579                                  American Hardcore   
4956              Mega Armageddon Death Extented "Live"   
4967                             Assimilated Pollutants   
5063                                       Heavy Crowns   
5133                           You're the New Metalcore   

                     artist                                   uri  \
145   Agoraphobic Nosebleed  spotify:track:4ESpRPaf5sBc4rEPCkKtba   
255              Enno Poppe  spotify:track:1krOnVXLfg0PJvipYsQ0iB   
631           Parts & Labor  spotify:track:6vSwfQIcybCa3FUZFSZvZC   
666          Insect Warfare  spotify:track:0pbqL0YsmbwdzRdEVRg7bK   
1595          Parts & Labor  spotify:track:5sttty8XOFJ188qopoVnGk   
2126      Few Left Standing  spotify:track:2CRcS8luNd3Is5KzTVKzil   
2222       Tears of Mankind  spotify:track:6ig3b62lNs9NdnBXGTaZyY   
3431         Carnifull Trio  spotify:track:2UvgupGVaCx11Xzd9TQt7J   
3621      Throwing Toasters  spotify:track:1qUnuTDWPZTmeJtUwNQqII   
4579         Weekend Nachos  spotify:track:1hIQKDecLtBLHBeGz22Upw   
4956        Electro Hippies  spotify:track:3Y2MxAyeOfiJoljr3yUwOc   
4967          Magrudergrind  spotify:track:77cX2pT9y8htWPa5lGH6XZ   
5063         Half the World  spotify:track:7ePR1sCX5GUfmRnPPCzrYk   
5133             Tiefighter  spotify:track:44BOMtl7BjAGukyoktM21m   

      danceability  energy  key  loudness  mode  speechiness  acousticness  \
145         0.2980   0.943   10    -4.019     0       0.1800      0.000085   
255         0.3360   0.545    1   -16.639     1       0.0981      0.789000   
631         0.0858   0.929    6    -5.755     1       0.1380      0.000745   
666         0.2220   0.994    1    -5.394     1       0.1250      0.000107   
1595        0.2540   0.786    7    -4.252     1       0.0605      0.000002   
2126        0.3680   0.900    6    -9.030     0       0.0599      0.001800   
2222        0.1720   0.701    3   -10.812     1       0.0376      0.013900   
3431        0.5730   0.453    2   -15.017     1       0.8560      0.993000   
3621        0.5110   0.415    2   -11.279     1       0.8220      0.894000   
4579        0.2760   0.796    1    -4.841     1       0.0873      0.137000   
4956        0.3530   0.670    2   -12.232     1       0.3140      0.692000   
4967        0.2300   0.969    2    -5.409     1       0.1450      0.000002   
5063        0.5550   0.860    3    -5.930     0       0.0795      0.000791   
5133        0.2110   0.653    2    -3.985     1       0.0772      0.000174   

      instrumentalness  liveness  valence    tempo  duration_ms  \
145           0.003820    0.5200   0.1990  102.700        34560   
255           0.000000    0.4300   0.2650   86.013        31787   
631           0.834000    0.5700   0.1200  169.809        28920   
666           0.000408    0.1810   0.0247   66.222        53453   
1595          0.985000    0.4250   0.1110  110.339        55427   
2126          0.889000    0.2700   0.3460  138.358        27533   
2222          0.909000    0.0412   0.1860  162.364       648467   
3431          0.347000    0.0700   0.3080  101.450        38348   
3621          0.000000    0.9650   0.5200   99.449        23133   
4579          0.006060    0.2330   0.0853   87.101        25880   
4956          0.000036    0.9740   0.0910  122.955        24107   
4967          0.933000    0.3340   0.2260  137.411        42733   
5063          0.001100    0.3620   0.9630   75.969        42721   
5133          0.000

In [ ]:
#some duplicate URIs
#some of these are slightly different versions of the same track by the same artist...some of them are different songs by the same artist...
duplicateURI = df[df.duplicated(['uri'], keep=False)].sort_values(by='uri')
duplicateURI

track  \
5302                                      Hey Ma   
4619                                      Hey Ma   
5828                                I Don't Care   
4836           I Don't Care (feat. Adam Gontier)   
5300                                     Want To   
356                             All I Want To Do   
4077                       It Just Comes Natural   
1835                                Give It Away   
5389                                      Shawty   
1088             Buy U A Drank (Shawty Snappin')   
5275                          If I Ain't Got You   
5657                                       Diary   
473                               Because Of You   
916                               Because Of You   
1515                                  South Side   
1496                                  South Side   
3105                             You Raise Me Up   
2612                             You Raise Me Up   
2627                                        I Am   
1281                             Take Me As I Am   
425                                  Mockingbird   
5683                                      Encore   
1552                          Here Comes Goodbye   
4016                                        Here   
1880                                No One Knows   
4896                                No One Knows   
1058                      Rock Wit U (Awww Baby)   
793                                         Baby   
410                         When Love Takes Over   
1330  When Love Takes Over (feat. Kelly Rowland)   
1859                Jai Ho! (You Are My Destiny)   
3706                Jai Ho! (You Are My Destiny)   
4193                                     Goodies   
5652                                   1, 2 Step   

                                                 artist  \
5302  Cam'Ron Featuring Juelz Santana, Freekey Zekey...   
4619                                            Cam’ron   
5828                Apocalyptica Featuring Adam Gontier   
4836                                       Apocalyptica   
5300                                          Sugarland   
356                                           Sugarland   
4077                                      George Strait   
1835                                      George Strait   
5389                             Plies Featuring T-Pain   
1088                          T-Pain Featuring Yung Joc   
5275                                        Alicia Keys   
5657            Alicia Keys Featuring Tony! Toni! Tone!   
473                                      Kelly Clarkson   
916              Reba McEntire Duet With Kelly Clarkson   
1515                        Moby Featuring Gwen Stefani   
1496                                               Moby   
3105                                        Josh Groban   
2612         Josh Groban & The African Children's Choir   
2627                                      Mary J. Blige   
1281                                      Mary J. Blige   
425                                              Eminem   
5683                          Eminem, Dr. Dre & 50 Cent   
1552                                      Rascal Flatts   
4016                                      Rascal Flatts   
1880                            Queens Of The Stone Age   
4896                            Queens of the Stone Age   
1058                                            Ashanti   
793                                             Ashanti   
410                David Guetta Featuring Kelly Rowland   
1330                                       David Guetta   
1859  A R Rahman & The Pussycat Dolls Featuring Nico...   
3706                                        A.R. Rahman   
4193                        Ciara Featuring Petey Pablo   
5652                      Ciara Featuring Missy Elliott   

                                       uri  danceability  energy  key  \
5302  spotify:track:0Uhnzk5zI46IRlQ04LNOtc         0.791   0.666    8   
4619  spot

In [ ]:
hitList = df.loc[df["target"]==1]
most_popular_artists = hitList["artist"].value_counts()
most_popular_artists.head(20)

Toby Keith           27
Tim McGraw           24
Rascal Flatts        24
Kenny Chesney        23
George Strait        22
Brad Paisley         20
Keith Urban          20
Alan Jackson         19
Britney Spears       19
Brooks & Dunn        17
Taylor Swift         17
P!nk                 15
Lonestar             15
Montgomery Gentry    15
Eminem               14
Alicia Keys          13
Nickelback           13
Martina McBride      13
Trace Adkins         13
JAY-Z                13
Name: artist, dtype: int64

In [ ]:
df.loc[(df['duration_ms']>600000) & (df['target']==1)]

track    artist                                   uri  \
859  Trapped In The Closet  R. Kelly  spotify:track:45LrQ3tg1z8plpuQRCuSwE   

     danceability  energy  key  loudness  mode  speechiness  acousticness  \
859         0.551   0.583    4    -6.473     1        0.159         0.539   

     instrumentalness  liveness  valence    tempo  duration_ms  \
859               0.0     0.365    0.419  107.424       992160   

     time_signature  chorus_hit  sections  target  
859               4     22.0433        29       1

In [1]:
df.loc[df['loudness'] > 1.0, 'loudness'] = 1.0

NameError: ignored